### PR TITLE
Allow 0-dim batch sizes for AdaptiveMaxPool and MaxPool.

### PR DESCRIPTION
**Pull request by v0dro**
This issue fixes a part of https://github.com/pytorch/pytorch/issues/12013, which is summarized concretely in  https://github.com/pytorch/pytorch/issues/38115.

This PR allows `MaxPool` and `AdaptiveMaxPool` to accept tensors whose batch size is 0. Some changes have been made to modernize the tests so that they will show the name of C++ function that throws an error.

### CODE REVIEW

**Comment by v0dro** [comment]
@jbschlosser could you please review this PR?

File: aten/src/ATen/native/AdaptiveMaxPooling2d.cpp [hunk 1]
```diff
@@ -6,18 +6,20 @@
...
-        "adaptive_max_pool2d: expected input to have non-empty spatial dimensions, "
+        "adaptive_max_pool2d(): Expected input to have non-empty spatial dimensions, "
         "but input has sizes ", input.sizes(), " with dimension ", i,
         " being empty");
   }
 
-  TORCH_CHECK((input.ndimension() == 3 || input.ndimension() == 4),
-      "non-empty 3D or 4D (batch mode) tensor expected for input");
+  TORCH_CHECK(ndim == 3 || ndim == 4,
+    "adaptive_max_pool2d(): Expected non-empty 3D or 4D tensor with optional 0-dim batch size, but got: ",
```
Comment (jbschlosser) [inline aten/src/ATen/native/AdaptiveMaxPooling2d.cpp:18]: nit: why not just "Expected 3D or 3D input tensor, but got:" ?

"non-empty" isn't fully accurate now, and I don't think "optional 0-dim batch size" needs to be specified

Same for error messages throughout the PR

**Review by jbschlosser — COMMENTED** [review]
Thanks for the update! Changes look pretty good- just one request below regarding error messages. I think it's best to be consistent with terminology throughout, and I think supporting "zero batch size" most accurately describes the goal (i.e. a size of 0 is allowed for the batch dim). Note that this is different from #60585, which is orthogonal and relates to leaving out the batch dim entirely.

File: aten/src/ATen/native/AdaptiveMaxPooling2d.cpp [hunk 2]
```diff
@@ -6,18 +6,20 @@
...
-        "adaptive_max_pool2d: expected input to have non-empty spatial dimensions, "
+        "adaptive_max_pool2d(): Expected input to have non-zero size for non-batch dimensions, "
         "but input has sizes ", input.sizes(), " with dimension ", i,
         " being empty");
   }
 
-  TORCH_CHECK((input.ndimension() == 3 || input.ndimension() == 4),
-      "non-empty 3D or 4D (batch mode) tensor expected for input");
+  TORCH_CHECK(ndim == 3 || ndim == 4,
+    "adaptive_max_pool2d(): Expected 3D or 4D tensor with optional 0-dim batch size, but got: ",
```
Comment (jbschlosser) [inline aten/src/ATen/native/AdaptiveMaxPooling2d.cpp:18]: nit: I think since this check is only concerned with the number of dims, "adaptive_max_pool2d(): Expected 3D or 4D tensor, but got: " is better

File: aten/src/ATen/native/AdaptiveMaxPooling3d.cpp [hunk 3]
```diff
@@ -7,10 +7,15 @@
 namespace at {
 namespace meta {
 TORCH_META_FUNC(adaptive_max_pool3d) (const Tensor& input, IntArrayRef output_size) {
-  for (int64_t i = 0; i < input.ndimension(); i++) {
+  auto ndim = input.ndimension();
+  TORCH_CHECK(
+    ndim == 4 || ndim == 5,
+    "adaptive_max_pool3d(): Expected 4D or 5D tensor with optional 0-dim batch size, but got: ",
```
Comment (jbschlosser) [inline aten/src/ATen/native/AdaptiveMaxPooling3d.cpp:13]: nit: I think since this check is only concerned with the number of dims, "adaptive_max_pool3d(): Expected 4D or 5D tensor, but got: " is better

File: aten/src/ATen/native/Pool.h [hunk 4]
```diff
@@ -205,8 +206,15 @@ pool3d_shape_check(
               "dilation should be greater than zero, but got ",
               "dilationT: ", dilationT, " dilationH: ", dilationH, " dilationW: ", dilationW);
 
-  TORCH_CHECK(input.numel() > 0 && (ndim == 4 || ndim == 5),
-              "non-empty 4D or 5D (batch mode) tensor expected for input, but got ndim: ", ndim);
+  TORCH_CHECK(ndim == 4 || ndim == 5,
+              fn_name, ": Expected 4D or 5D tensor for input with optional non-zero batch dim, but got ndim: ",
```
Comment (jbschlosser) [inline aten/src/ATen/native/Pool.h:210]: nit: I think since this check is only concerned with the number of dims, "Expected 4D or 5D input tensor, but got ndim: " is better

**Review by jbschlosser — COMMENTED** [review]
Some tiny minor comments then good to go! :)

**Review by jbschlosser — APPROVED** [review]
LGTM! Thanks for the update :)

@v0dro FYI the test failures are real; `test_max_pool1d_errors` in `test/test_nn.py` should be updated to reflect the error message changes.

File: test/test_nn.py [hunk 5]
```diff
@@ -13840,8 +13891,8 @@ def check(x, args, message):
                 model(torch.tensor(x, device=device, dtype=dtype))
 
         # Pooling args: (kernel_size, stride, padding, dilation, return_indices, ceil_mode)
-        check(0, (1,), "input tensor must have 2 or 3 dimensions but got 0")
-        check([], (1,), "input tensor must have 2 or 3 dimensions but got 1")
+        check(0, (1,), "Expected 2D or 3D input tensor, but got 0")
```
Comment (jbschlosser) [inline test/test_nn.py:13894]: Check is still failing, I think changing the error message to "Expected 2D or 3D input tensor, but got" without the 0 should do the trick.

File: test/test_nn.py [hunk 6]
```diff
@@ -13840,8 +13891,8 @@ def check(x, args, message):
...
+        check([], (1,), "Expected 2D or 3D input tensor, but got 1")
```
Comment (jbschlosser) [inline test/test_nn.py:13895]: Similarly here, removing the " 1" should get the check passing.

Comment (v0dro) [reply test/test_nn.py]: Extremely sorry for the mistake. Have updated it.
